Render accented characters

lmondada opened the pull request:
Add support for rendering of accented chars in `RenderParsedTextResolver`. Previously, the `Render` function for `TextRenderer` would fail with a `renderer - unhandled ast.Expr type` error when trying to render a `TextAccent`.

Summary of changes:
- The feature is implemented in the `renderer.go` file. If you think the logic is too verbose, this could also be moved into a new `accents.go` file.
- In the modified `Render` function in `renderer.go`, fixed a minor bug in how errors are propagated. Previously, the function terminated with `return nil`, which would not return the `mErr` error, as some of the `switch` branches were expecting.
- In the `Resolve` function of `resolver.go`, removed a `for` loop that was writing to an `expr` value that was no longer in the AST, and thus these modifications were ignored. Everything is already properly taken care of within `Render`.
- In `TestRenderParsedTextResolver_Resolve` of `resolver_test.go`, pass nodes wrapped within `TagStmt` to `RenderParsedTextResolver`. Otherwise the walker within `Resolve` ignores the node.

These issues popped up as I was trying to add a test for my case. It ended up touching more code than I intended originally.

Let me know if I've read anything wrong.

lmondada committed:
Render accented characters

jschaf commented on render/render.go:
```diff
@@ -240,18 +243,137 @@ func (p TextRenderer) Render(w io.Writer, x ast.Expr) (mErr error) {
+	baseChar = unicode.ToLower(baseChar)
+
+	// Mapping of base characters to their accented versions
+	accentMap := map[accentType]map[rune]rune{
```
The go compiler might be smart enough to hoist the map, but let's move it outside the function since it's static.

jschaf commented on render/render.go:
```diff
@@ -240,18 +243,137 @@ func (p TextRenderer) Render(w io.Writer, x ast.Expr) (mErr error) {
+	}
+
+	// Check if the accent exists for the given base character
+	if accentedChars, exists := accentMap[accent]; exists {
```
Can we keep the happy path unindented, i.e. https://maelvls.dev/go-happy-line-of-sight/:

```go
accentedChars, exists := accentMap[accent]
if ! exists {
    return fmt.Errorf(...)
}
accented, valid := accentedChars[baseChar]
if !valid {...}

return accented, nil
```

lmondada replied:
Oh, this is a nice rule I will apply outside of Go! Thanks for the pointer.

jschaf commented on render/render.go:
```diff
@@ -240,18 +243,137 @@ func (p TextRenderer) Render(w io.Writer, x ast.Expr) (mErr error) {
+			'a': 'à', 'e': 'è', 'i': 'ì', 'o': 'ò', 'u': 'ù',
+			'A': 'À', 'E': 'È', 'I': 'Ì', 'O': 'Ò', 'U': 'Ù',
+		},
+		Acute: {
```
It looks like the reason for the two-layer map is to generate helpful error messages. I think we can accomplish the same thing with a plain map.

```go
accentMap := map[string]string{
  "`a": "à",
  ".c": "ċ",
}
```

For the error message, scan the map by prefix.

jschaf commented on render/render.go:
```diff
@@ -211,8 +214,8 @@ func (p TextRenderer) Render(w io.Writer, x ast.Expr) (mErr error) {
 		for _, v := range t.Values {
-			if err := p.Render(w, v); err != nil {
-				return err
+			if mErr = p.Render(w, v); mErr != nil {
```
mErr is a convention I use for capturing deferred errors. Since we don't use it here, let's remove mErr entirely and just use `err`.

https://news.ycombinator.com/item?id=41479891

lmondada replied:
ok, great, thanks for the pointer. I interpreted this as: use `mErr` when you don't return immediately, and `err` otherwise.

I've adjusted the code accordingly, let me know if you'd like me to remove all `mErr` uses instead.

jschaf commented on render/render.go:
```diff
@@ -240,18 +243,137 @@ func (p TextRenderer) Render(w io.Writer, x ast.Expr) (mErr error) {
+}
+
+// accentType represents the different LaTeX accents
+type accentType string
```
Let's move these to accent.go. It turns out there's a lot of accented characters we might want to eventually support: https://github.com/loopspace/latexToUTF/blob/master/utf.php

lmondada replied:
done!

jschaf commented on render/render.go:
```diff
@@ -240,18 +243,137 @@ func (p TextRenderer) Render(w io.Writer, x ast.Expr) (mErr error) {
+		"`":  Grave,
+		"'":  Acute,
+		"^":  Circumflex,
+		"\"": Umlaut,
```
optional: backticks would let you avoid the backslash

jschaf commented on render/render.go:
```diff
@@ -240,18 +243,137 @@ func (p TextRenderer) Render(w io.Writer, x ast.Expr) (mErr error) {
-	return nil
+
+	// Look up the accent type
+	if accentType, exists := accentMap[accentStr]; exists {
```
Similar to the other comment, let's unindent the happy path.

jschaf commented on render/render.go:
```diff
@@ -240,18 +243,137 @@ func (p TextRenderer) Render(w io.Writer, x ast.Expr) (mErr error) {
+	accentStr = strings.ToLower(strings.TrimSpace(accentStr))
+
+	// Map of string representations to AccentType
+	accentMap := map[string]accentType{
```
hoist the map out of the function.

jschaf commented on render/render.go:
```diff
@@ -240,18 +243,137 @@ func (p TextRenderer) Render(w io.Writer, x ast.Expr) (mErr error) {
+		if utf8.RuneCountInString(text) != 1 {
+			return fmt.Errorf("renderer - accent text must be a single character")
+		}
+		r, _ := utf8.DecodeRuneInString(text)
```
Is the value always a rune? If so, let's change the type to rune instead of ast.Text.

lmondada replied:
I've looked for a definitive answer, and it does seem that it [always a single rune](https://en.wikibooks.org/wiki/LaTeX/Special_Characters#Escaped_codes).

However, checking LaTeX's behaviour: the string `\'{ee}` is valid and parsed as `\'ee`. I've thus left the accented text to be a string and modified the rendering logic to leave the characters past the first one unchanged.

jschaf reviewed (changes requested):
Looks pretty good. Few stylistic comments and a change to the ast.TextAccent type.

lmondada committed:
Address comments

lmondada committed:
Move accent rendering logic to accent.go

lmondada committed:
Handle mulit-rune accents

lmondada committed:
Change mErr -> err when returned immediately

lmondada committed:
enhance test name

lmondada commented:
Thanks @jschaf for the good review and useful pointers. All your points should be addressed!

Sorry it took a while!

jschaf committed:
Polish

Make ast.TextAccent use a rune token and ast.Text value.

Move selecting the accent rune into the scanner since that's the
earliest spot we can do it.

jschaf commented on ast/ast.go:
```diff
@@ -203,9 +203,9 @@ type (
-		Value    Expr      // the text to accent
+		ValuePos gotok.Pos    // literal position
+		Accent   token.Accent // the accent character, like "^" for circumflex
+		Text     *Text        // the text to accent; must be a single rune
```
The parser requires this to be ast.Text, so change the type to simplify life when traversing the AST to render it.

jschaf commented on scanner/scanner.go:
```diff
@@ -307,7 +307,13 @@ func (s *Scanner) scanStringEscape() (token.Token, string) {
 		s.next()
 		return token.StringBackslash, string(s.src[offs:s.offset])
-	case '\'', '~', '.', '^', '=', '`', '"':
+	case rune(token.AccentAcute),
```
Make sure the scanner uses the same set of supported accent chars.

jschaf reviewed (approved):
lgtm! I did a quick polish. I'll merge it shortly.
